### PR TITLE
posix: add support for mmap, memlock, shared memory, and mprotect

### LINKED ISSUE #59950: posix: implement `mmap()`
There are currently 20 missing POSIX functions with trivial implementations, `mmap()` is one of them.

> mmap - map pages of memory
> 
> The mmap() function shall establish a mapping between an address space of a process and a memory object.
> ...

See https://pubs.opengroup.org/onlinepubs/9699919799/functions/mmap.html

See RFC #51211 for more info.

### LINKED ISSUE #59951: posix: implement `munmap()`
There are currently 20 missing POSIX functions with trivial implementations, `munmap()` is one of them.

> munmap - unmap pages of memory
> 
> The munmap() function shall remove any mappings for those entire pages containing any part of the address space of the process starting at addr and continuing for len bytes. Further references to these pages shall result in the generation of a SIGSEGV signal to the process. If there are no mappings in the specified address range, then munmap() has no effect.
> 
> The implementation may require that addr be a multiple of the page size as returned by [sysconf()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sysconf.html).
> 
> If a mapping to be removed was private, any modifications made in this address range shall be discarded.
> 
> [[ML|MLR](javascript:open_code('ML'))] [Option Start] Any memory locks (see [mlock](https://pubs.opengroup.org/onlinepubs/9699919799/functions/mlock.html#) and [mlockall](https://pubs.opengroup.org/onlinepubs/9699919799/functions/mlockall.html#)) associated with this address range shall be removed, as if by an appropriate call to [munlock()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/munlock.html). [Option End]
> 
> [[TYM](javascript:open_code('TYM'))] [Option Start] If a mapping removed from a typed memory object causes the corresponding address range of the memory pool to be inaccessible by any process in the system except through allocatable mappings (that is, mappings of typed memory objects opened with the POSIX_TYPED_MEM_MAP_ALLOCATABLE flag), then that range of the memory pool shall become deallocated and may become available to satisfy future typed memory allocation requests.
> 
> A mapping removed from a typed memory object opened with the POSIX_TYPED_MEM_MAP_ALLOCATABLE flag shall not affect in any way the availability of that typed memory for allocation. [Option End]
> 
> The behavior of this function is unspecified if the mapping was not established by a call to [mmap()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/mmap.html).

See https://pubs.opengroup.org/onlinepubs/9699919799/functions/munmap.html

See RFC #51211 for more info.

### LINKED ISSUE #59952: posix: implement `shm_open()`
There are currently 20 missing POSIX functions with trivial implementations, `shm_open()` is one of them.

> shm_open - open a shared memory object (REALTIME)
> 
> The shm_open() function shall establish a connection between a shared memory object and a file descriptor. It shall create an open file description that refers to the shared memory object and a file descriptor that refers to that open file description. The file descriptor shall be allocated as described in [File Descriptor Allocation](https://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_14), and can be used by other functions to refer to that shared memory object. The name argument points to a string naming a shared memory object. It is unspecified whether the name appears in the file system and is visible to other functions that take pathnames as arguments. The name argument conforms to the construction rules for a pathname, except that the interpretation of <slash> characters other than the leading <slash> character in name is implementation-defined, and that the length limits for the name argument are implementation-defined and need not be the same as the pathname limits {PATH_MAX} and {NAME_MAX}. If name begins with the <slash> character, then processes calling shm_open() with the same value of name refer to the same shared memory object, as long as that name has not been removed. If name does not begin with the <slash> character, the effect is implementation-defined.
> ...

See https://pubs.opengroup.org/onlinepubs/9699919799/functions/shm_open.html

See RFC #51211 for more info.

### LINKED ISSUE #59953: posix: implement `shm_unlink()`
There are currently 20 missing POSIX functions with trivial implementations, `shm_unlink()` is one of them.

> shm_unlink - remove a shared memory object (REALTIME)
> 
> The shm_unlink() function shall remove the name of the shared memory object named by the string pointed to by name.
> 
> If one or more references to the shared memory object exist when the object is unlinked, the name shall be removed before shm_unlink() returns, but the removal of the memory object contents shall be postponed until all open and map references to the shared memory object have been removed.
> 
> Even if the object continues to exist after the last shm_unlink(), reuse of the name shall subsequently cause [shm_open()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/shm_open.html) to behave as if no shared memory object of this name exists (that is, [shm_open()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/shm_open.html) will fail if O_CREAT is not set, or will create a new shared memory object if O_CREAT is set).

See https://pubs.opengroup.org/onlinepubs/9699919799/functions/shm_unlink.html

See RFC #51211 for more info.

### PR DESCRIPTION
**Pull request by cfriedt**
* add support for `_POSIX_MAPPED_FILES`
* add support for `_POSIX_MEMLOCK_RANGE`
* add support for `_POSIX_MEMLOCK` (currently unimplemented / fails with `ENOSYS`)
* add support for `_POSIX_SHARED_MEMORY_OBJECTS`
* add support for `_POSIX_MEMORY_PROTECTION` (currently unimplemented / fails with `ENOSYS`)

**Shared Memory Objects and Memory-Mapped Files**

On systems that support mapping anonymous memory (i.e. have an MMU and lots of RAM), shared memory objects are implemented using `k_mem_map()` and friends (i.e. a shared memory object is comprised of a number of blocks of virtual memory that are `CONFIG_MMU_PAGE_SIZE` in size). On systems that do not have an MMU, shared memory objects are implemented with heap allocation.

Note: it is up to the particular vtable implementor, to support the `ZFD_IOCTL_MMAP` operation. However, there would still need to be some integration with `msync()` in order to synchronize writes to e.g. a regular file. From that perspective, shared memory objects were an easy prototyping option.

A potential future improvement on shared memory objects would be to support a separate `ioctl()` to attach statically allocated memory.

**z_finalize_typed_fd()**

A particular change to highlight in this case though, is adding a `mode` field to `struct fd_entry`, which allows fdtable code to identify which subsystem or API a file descriptor belongs to, and to react accordingly. The `z_finalize_typed_fd()` function allows us to differentiate between fifo, char device, message queue, directory, semaphore, block device, shared memory, regular file, symbolic link, and socket.

In fact, the same 32-bit `mode` field _should_ also have the file access mode tied to it, rather than the physical file (which is currently the case now for the FS subsystem, and should be considered a bug). It's trivial to consider a case where one file descriptor is used for writing to a file and another file descriptor is used for reading from a file. Both file descriptors point to the same physical file, but each has different access modes and offsets (addressed below).

**Per-file Offsets**

Another similar adjustment to `struct fd_entry` is to have a per-fd (per `FILE*`) offset, which is just consistent with the C/C++, POSIX (all?) programming models of "files".

A few drive-by fixes aside from that, shuffling things between zvfs, posix, and fs, as-needed.

[Doc Preview](https://builds.zephyrproject.io/zephyr/pr/73799/docs/services/portability/posix/index.html)

Fixes #59950
Fixes #59951
Fixes #59952
Fixes #59953

### CODE REVIEW

**Comment by cfriedt** [comment]
Some platforms seem to have `CONFIG_MMU` but do not implement `arch_mem_map()` or `arch_mem_unmap()`, so there may be a couple of exclusions in `tests/posix/shm/testcase.yaml`

File: lib/posix/options/mlock.c [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <stddef.h>
+#include <sys/types.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/posix/sys/mman.h>
```
Comment (ycsin) [inline lib/posix/options/mlock.c:12]: nit: we probably have more headers here than we need

File: lib/posix/options/mmap.c [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	return addr;
+}
+
+int msync(void *addr, size_t length, int flags)
+{
+	ARG_UNUSED(addr);
+	ARG_UNUSED(length);
+	ARG_UNUSED(flags);
+
+	return 0;
```
Comment (ycsin) [inline lib/posix/options/mmap.c:92]: so this will just pass silently? should we return -1 (& ENOSYS) instead?

**Review by ycsin — COMMENTED** [review]
Skimmed through and looks reasonable to me, just some nits

File: lib/posix/options/mmap.c (hunk 2 above)
Comment (cfriedt) [reply lib/posix/options/mmap.c:92]: Currently, it doesn't seem that there is a generic API to flush anonymously mapped pages. Also from what I understand, the memories used in this implementation so far use synchronized writes. I could double-check that though.

Eventually, when mapping regular files is implemented by the fs subsystem, there will need to be an ISR to catch writes, and those might also be buffered, but that's a future problem.

Otherwise, it may be useful to integrate the cache API eventually, but it's probably also premature at this point.

https://docs.zephyrproject.org/latest/hardware/cache/index.html

Comment (cfriedt) [reply lib/posix/options/mmap.c:92]: Yeah, actually, so `k_mem_map_impl()` does not support explicit caching flags, so all writes will be synchronous. There is no need to fail, since doing nothing and succeeding has the expected result.

https://github.com/zephyrproject-rtos/zephyr/blob/453ab8a9a356acf475a965a777a370795effa255/kernel/mmu.c#L581

**Review by ycsin — DISMISSED** [review]

File: lib/posix/options/mmap.c [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+{
+	int zflags = p2z(prot, flags);
+
+	if ((len == 0) || (zflags == -1)) {
+		errno = EINVAL;
+		return MAP_FAILED;
+	}
+
+	if (fd > 0) {
+		/* non-anonymous mapping */
```
Comment (andyross) [inline lib/posix/options/mmap.c:58]: Driving by, haven't had time for a full review but this all looks pretty straightforward:

Is MAP_ANON/MAP_ANONYMOUS not part of POSIX?  You probably want that here for portability, since that's the primary mechanism in linux.  Linux will ignore the fd argument if it's set (it doesn't have to be -1).  This code will use the fd only, meaning that Linux code that gets ported to it (i.e. drops the MAP_ANONYMOUS flag) and forgets to pass -1 in the fd argument (because it was never needed) is going to misbehave.

Comment (cfriedt) [reply lib/posix/options/mmap.c:58]: > Driving by, haven't had time for a full review but this all looks pretty straightforward:
> 
> Is MAP_ANON/MAP_ANONYMOUS not part of POSIX? 

It is (maybe) surprisingly not POSIX.

> You probably want that here for portability, since that's the primary mechanism in linux. 

Probably, yeah.

> Linux will ignore the fd argument if it's set (it doesn't have to be -1).  This code will use the fd only, meaning that Linux code that gets ported to it (i.e. drops the MAP_ANONYMOUS flag) and forgets to pass -1 in the fd argument (because it was never needed) is going to misbehave.

I was thinking something along the line of forcing the fd argument to -1 if the `MAP_ANONYMOUS` flag is present. That should work, right?

Since this PR is already holding up quite a few more, would it be acceptable to submit that change as a follow up, @andyross?

Comment (andyross) [reply lib/posix/options/mmap.c:58]: Oh sure; it was just a note. I try to be pretty clear with my -1's. :)

**Review by vaishnavachath — DISMISSED** [review]

**Comment by npitre** [comment]
Here's the worst offenders I've encountered.

In `munmap()` you shouldn't call `arch_page_phys_get()` as this will
return -EFAULT if the first page is paged out. Should call `k_mem_unmap()`
directly.

Side note: I think `k_mem_unmap()` should return an error code instead of
void. May always be 0 for now but the prototype should be ready.

`shm_mmap()` looks interesting. This:

- `arch_page_phys_get(shm->mem, &phys)`

followed by:

- `*virt = k_mem_phys_map(phys + off, ROUND_UP(len, _page_size), K_MEM_PERM_RW)`

is going to produce very surprizing results with demand paging indeed.
First, because you use `k_mem_phys_map()` you are bypassing the very tiny
protection against having a physical page mapped twice. It is kinda OK only
with pinned memory which is another way to say that demand paging doesn't
support it.

In your case, what would happen is that you get a second mapping at a 
different virtual address but one that is not tracked by the physical page 
frame management code. So if memory pressure increases and `shm->mem` is not
accessed _irrespective_ of how often that second virtual mapping is used then
the `shm->mem` virtual address will be a prime candidate to be paged out.
And when that happens, the underlying physical page will be recycled for
some other content accessible through some other virtual address that now
will also be visible through the mapping you created with that
`k_mem_phys_map()`, and such content no longer having anything to do with 
the initial `shm->mem` content. And yet, if `shm->mem` is accessed again,
it is unlikely that its content will be paged back into the same physical
page.

In short, this is no good!  ;-)

Now... why do you need a second mapping? Zephyr uses a flat address space
model shared by all threads. Just return `shm->mem` from shm_mmap()`, turn
`shm_unmmap()` into a no-op and be done with it.

Eventually it would make sense to leverage memory domains if this is going
to be used in user space so that a shared memory block is part of the same 
memory protection domain.

**Review by dkalowsk — DISMISSED** [review]

**Review by npitre — CHANGES_REQUESTED** [review]
Just to make it obvious from my previous comment.

File: lib/posix/options/mlock.c (hunk 1 above)
Comment (cfriedt) [reply lib/posix/options/mlock.c:12]: Fixed!

File: lib/posix/options/mmap.c (hunk 3 above)
Comment (cfriedt) [reply lib/posix/options/mmap.c:63]: Fixed!

**Comment by cfriedt** [comment]
- [x] remove unneeded headers from `mlock.c` (requested by @ycsin)
- [x] add a compatible `MAP_ANONYMOUS` to force an anonymous mapping in `mmap.c` (requested by @andyross)
- [ ] attempted to call `k_mem_unmap()` directly in `mmap.c`, but that triggers an assert in `mmu.c`, so kept it for now
- [x] use `shm->mem` for all file descriptors and 1 virtual mapping shared among file descriptors (request by @npitre)

@npitre - PTAL 🙏 

ping @andyross @ycsin @dkalowsk @vaishnavachath for refresh +1 🙏

File: lib/os/fdtable.c [hunk 4]
```diff
@@ -373,9 +373,7 @@ int ioctl(int fd, unsigned long request, ...)
 	va_list args;
 	int res;
 
-	if (_check_fd(fd) < 0) {
-		return -1;
-	}
+	__ASSERT_NO_MSG(fd < ARRAY_SIZE(fdtable));
```
[the remark this answers is not in the document]
Comment (cfriedt) [reply lib/os/fdtable.c:400]: Fixed!

File: include/zephyr/sys/fdtable.h [hunk 5]
```diff
@@ -21,8 +35,14 @@ extern "C" {
  * Currently all operations beyond read/write/close go thru ioctl method.
  */
 struct fd_op_vtable {
-	ssize_t (*read)(void *obj, void *buf, size_t sz);
-	ssize_t (*write)(void *obj, const void *buf, size_t sz);
+	union {
+		ssize_t (*read)(void *obj, void *buf, size_t sz);
```
Comment (d3zd3z) [inline include/zephyr/sys/fdtable.h:39]: I have a little concern about the unions here resulting in some difficult to find bugs for future implementers. I guess the other option would be to have these separate, but that would take additional space, but is that a lot?

**Review by d3zd3z — COMMENTED** [review]
Other than this minor issue, it looks ok to me.

File: include/zephyr/sys/fdtable.h (hunk 5 above)
Comment (cfriedt) [reply include/zephyr/sys/fdtable.h:39]: Pushing an extra, unused, argument onto the stack and popping it back off shouldn't be too much of a concern I would think. But up to 16 extra (unneeded / unused) bytes per file descriptor might be a lot of wasted space, especially given that most seekable file types should eventually adopt the `_offset()` variants, and it's they are otherwise identical.

It should be noted as well, that offsets do not work apply to sockets, pipes, etc, just regular files and memory.

**Comment by npitre** [comment]
In `shm_obj_remove()` you do:

```
                if (arch_page_phys_get(shm->mem, &phys) == 0) {
                        k_mem_unmap(shm->mem, ROUND_UP(shm->size, _page_size));
                }
```

I'd suggest removing that `arch_page_phys_get()`. It isn't useful there.
`k_mem_unmap()` should cope regardless.

And if the start address is paged out then `arch_page_phys_get()` will return
an error even if `k_mem_unmap()` still can unmap it (once #74044 is applied
that is).

**Comment by cfriedt** [comment]
> I'd suggest removing that `arch_page_phys_get()`. It isn't useful there. `k_mem_unmap()` should cope regardless.

Fixed!

> And if the start address is paged out then `arch_page_phys_get()` will return an error even if `k_mem_unmap()` still can unmap it (once #74044 is applied that is).

I rebased on your PR 😎 🪁

**Comment by cfriedt** [comment]
And then I rebased again on `main`! 😅

**Comment by npitre** [comment]
> > I'd suggest removing that `arch_page_phys_get()`. It isn't useful there.
> > `k_mem_unmap()` should cope regardless.
> 
> Fixed!

I do see my `k_mem_unmap()` changes ... but the `arch_page_phys_get()` call
is still there.

**Comment by cfriedt** [comment]
> > > I'd suggest removing that `arch_page_phys_get()`. It isn't useful there.
> > > `k_mem_unmap()` should cope regardless.
> > 
> > 
> > Fixed!
> 
> I do see my `k_mem_unmap()` changes ... but the `arch_page_phys_get()` call is still there.

@npitre - There seems to be some unwanted side-effects on rebase. I wonder if I could maybe just go back to the previous version that did the questionable check that was passing CI. Maybe the last concern of yours can be addressed in a bugfix?

**Comment by npitre** [comment]
> @npitre - There seems to be some unwanted side-effects on rebase.
> I wonder if I could maybe just go back to the previous version that
> did the questionable check that was passing CI. Maybe the last concern
> of yours can be addressed in a bugfix?

Fair enough.

**Comment by cfriedt** [comment]
> Fair enough.

@npitre - rebased with your change now that it's in `main`.

I needed to drop the assert in `k_mem_unmap_phys_guard()`, otherwise it would complain / die.

I think that's how you were saying `k_mem_unmap()` should behave anyway though, and from my perspective, I think we should be able to call `k_mem_unmap()` with a virtual address that might not be mapped, and it should return quietly 🤷‍♂️

**Review by ycsin — DISMISSED** [review]

File: include/zephyr/sys/fdtable.h (hunk 5 above)
Comment (d3zd3z) [reply include/zephyr/sys/fdtable.h:39]: I guess I was more concerned with the subtle semantic difference. If the implementation is an offset, but the caller is expecting without, that offset argument will be read as garbage.

Comment (cfriedt) [reply include/zephyr/sys/fdtable.h:39]: Interestingly, `lseek()` must be able to set the file descriptor offset to an arbitrary value. Kind of mind boggling.

> The lseek() function shall allow the file offset to be set beyond the end of the existing data in the file.

https://pubs.opengroup.org/onlinepubs/9699919799/functions/lseek.html

Consequently, read and write implementations must be able to decide if an offset is valid or invalid and react accordingly.

So they should be able to handle garbage.

**Review by npitre — DISMISSED** [review]

**Comment by nashif** [comment]
> z_finalize_typed_fd()

@cfriedt  
dang, we still have thise z_ prefixed public APIs, we should have addressed those long time ago :(
Any chance you can address this in a followup PR after this get merged?

**Comment by npitre** [comment]
> I needed to drop the assert in `k_mem_unmap_phys_guard()`, otherwise it
> would complain / die.
> 
> I think that's how you were saying `k_mem_unmap()` should behave anyway
> though, and from my perspective, I think we should be able to call
> `k_mem_unmap()` with a virtual address that might not be mapped, and it 
> should return quietly🤷♂️

Well... In _theory_ you are not supposed to unmap() the same memory twice.
So if that assertion still triggered then _you_ have a bug somewhere. ;-)

Granted, this is no reason for killing the system in production.
That's why assertions are then disabled. But it would be a good thing if
you found why you are calling unmap twice and then reinstating the
assertion.

My earlier comments were made in the context of demand paging where
`arch_page_phys_get()` may not be relied upon. But even then, it is not
OK to use it to dodge the assertion on a double unmap.

**Review by vaishnavachath — DISMISSED** [review]

File: kernel/mmu.c [hunk 6]
```diff
@@ -722,11 +722,8 @@ void k_mem_unmap_phys_guard(void *addr, size_t size, bool is_anon)
 #else
 			ret = arch_page_phys_get(pos, &phys);
 #endif
-			__ASSERT(ret == 0,
-				 "%s: cannot unmap an unmapped address %p",
-				 __func__, pos);
```
Comment (dcpleung) [inline kernel/mmu.c:727]: This is here to catch coding error. If a continuously mapped virtual memory space has an unmapped hole, this means someone who should not be poking at this region has unmapped some parts of it. The original owner/mapper of this region has no idea that there is a hole and will continue accessing it, resulting in unexpected page faults.

What exactly are you trying to solve here?

Comment (cfriedt) [reply kernel/mmu.c:727]: Originally I used some other call to check if a region was mapped. If the call failed, I knew not to call `k_mem_unmap()`. @npitre said not use the call to check if the region was mapped and to simply call `k_mem_unmap()` and that this function would simply return if the region was unmapped.

That is not the observed behaviour. The observed behaviour is that the assert is always triggered now.

There seems to be an impedance mismatch between how maintainers expect this function to work.

That impedance mismatch is unfortunately blocking quite a lot features for 3.7.0 have otherwise been ready to merge for quite a while now.

I really hate asking for an extension after feature freeze, but likely have no choice at this point.

**Comment by cfriedt** [comment]
> > z_finalize_typed_fd()
> 
> @cfriedt  
> dang, we still have those z_ prefixed public APIs, we should have addressed those long time ago :(
> Any chance you can address this in a followup PR after this get merged?

Sure - I was planning on using zvfs, but would you prefer k?

**Comment by cfriedt** [comment]
> you found why you are calling unmap twice 

I am calling unmap twice because you said I should not check first to see if the region is mapped. Previously I had a call to see if the region was mapped.

Aside from that, it would duplicate a lot of effort to keep track of all mapped pages in a second location not really part of the kernel or arch.

**Comment by cfriedt** [comment]
> it is not
> OK to use it to dodge the assertion on a double unmap.

But it is ok to for any random bit of code to crash the system by calling k_mem_unmap() with a garbage argument? That seems like an obvious gap in testing.

It's not the user's responsibility to manage system resources, it's the kernel's responsibility.

**Comment by dcpleung** [comment]
Another issue is that your change will make the function to bail out in the middle of the for loop, leaving already mapped memory after the failing one possibly forever mapped, and never released. A more appropriate place to bail would be where the guard pages are checked. Moreover, the callers actually have to make sure they are unmapping with the same parameters when the region is mapped.

For you to actually hit that assert, you have to pass the two checks for the guard pages first. That means something is horribly wrong for a page in middle of that region to be unmapped. Another reason might be that you are unmapping across multiple virtual regions at the same time. So the boundary guard pages checks would pass, but when you are looping through the pages, you encounter the guard pages in between. For example, region A and region B with guard pages G: `GAAAGGBBBG`. You trying to unmap the whole thing which means the assert would be raised on the middle `GG` guard pages.

Also, this is a very low level API, not to mention being a kernel internal one, one must exercise extreme caution when using it (and also the corresponding mapping function). If garbage arguments are passed to these functions, it is either a coding error, or that memory has been corrupted (e.g. stack overflow, or incorrect pointer de-reference, or cosmic rays). So bailing would be a wise choice in the later case, as correct operations of the whole system can no longer be guaranteed.

**Comment by cfriedt** [comment]
> Another issue is that your change ...

I'm fine dropping the mmu change, but please just understand this is how I was told to use the function call. It is broken for the particular use case I was told to use it for. I realize that it is low level / not extremely frequently used code, etc, so I knew that there would be rough edges. This is why I've marked mmap as experimental.

I'll revert back to check first using the `arch_` function, since there seems to be no way to otherwise unmap memory using the POSIX model without crashing the system (the POSIX model being that pages can be mapped to more than one virtual memory address, and the kernel doesn't panic when it's told to unmap an invalid address).

**Comment by cfriedt** [comment]
* switched back to checking if a page is mapped via `arch_page_phys_get()` prior to calling `k_mem_unmap()` due to lack of a better alternative

**Comment by cfriedt** [comment]
re-ping @andyross @vaishnavachath @ycsin @dkalowsk

**Review by ycsin — APPROVED** [review]

**Comment by npitre** [comment]
> It's not the user's responsibility to manage system resources, it's the
> kernel's responsibility.

Sorry, but the POSIX layer you're creating (and that's highly appreciated
by the way) is _not_ user code. You are managing many different resources
already on behalf of code calling the POSIX interface. That code is the user
not the interface.

Knowing how many times a particular object has been mapped and unmapped is 
the POSIX layer's responsibility, just like the refcounting you already do
elsewhere between `malloc()` and `free()`. You wouldn't free the same memory
twice, right? And you're not calling into the low-level heap allocator to test
if a particular address is still allocated before calling `free()` either.
The assertion in `k_mem_unmap()` is there to catch the equivalent of a
double-free, which is a bug and why it shouldn't be removed.

So it should be no different with `k_mem_map()` and `k_mem_unmap()`.
The first caller to increment the refcount does the map, subsequent callers 
only increment the refcount and get the same already mapped address, and the
last to decrement the count to 0 does the unmap.

Am I missing something that would prevents such implementation?

**Review by vaishnavachath — APPROVED** [review]

**Comment by nashif** [comment]
> > > z_finalize_typed_fd()
> > 
> > 
> > @cfriedt
> > dang, we still have those z_ prefixed public APIs, we should have addressed those long time ago :(
> > Any chance you can address this in a followup PR after this get merged?
> 
> @nashif - Sure - I was planning on using zvfs, but would you prefer k?

No `k_`, Ideally this should be the new prefix we discussed in the f2f (https://github.com/zephyrproject-rtos/zephyr/issues/64627).

Let me open a bug about this and bring the prefix discussion back so we can do something about it.

**Comment by nashif** [comment]
> No `k_`, Ideally this should be the new prefix we discussed in the f2f (#64627).
> 
> Let me open a bug about this and bring the prefix discussion back so we can do something about it.

actually you can use zvfs which is being used already, better than z_ for sure.

**Comment by cfriedt** [comment]
> > > > z_finalize_typed_fd()
> > > 
> > > 
> > > @cfriedt
> > > dang, we still have those z_ prefixed public APIs, we should have addressed those long time ago :(
> > > Any chance you can address this in a followup PR after this get merged?
> > 
> > 
> > @nashif - Sure - I was planning on using zvfs, but would you prefer k?
> 
> No `k_`, Ideally this should be the new prefix we discussed in the f2f (#64627).
> 
> Let me open a bug about this and bring the prefix discussion back so we can do something about it.

@nashif - I think a `k_` prefix is fine for things like `z_finalize_fd()`, etc, but for consistency, that should likely just be a wrapper around a `zvfs_` prefixed function (mutual dependency library of net, fs, posix, libc, etc, where common socket / fs / fd stuff was heading since 2022, #51211).

**Comment by nashif** [comment]
> I think a `k_` prefix is fine for things like `z_finalize_fd()`

`k_` should be reserved for kernel/ stuff only, let not mix things up.

**Review by dcpleung — APPROVED** [review]

**Review by andyross — APPROVED** [review]
Still haven't combed through as well as I'd like, but +1 to everything I saw earlier.  And it seems like this has excellent review coverage from others anyway

**Review by dkalowsk — APPROVED** [review]
Still thinking this is better than what we have today.  Reviewing I don't see anything obvious standing out.

**Comment by npitre** [comment]
> Am I missing something that would prevents such implementation?

Of course I am. `munmap()` doesn't pass a fd or any identifier other than
the actual address and length.

But you have another problem here. `k_mem_map_phys_bare()` should be paired
only with `k_mem_unmap_phys_bare()` otherwise bad things will happen.
I'd suggest you use `k_mem_map_phys_guard()` here instead, as it is
much closer to be compatible with `k_mem_unmap()` aka
`k_mem_unmap_phys_guard()`.

And yet `k_mem_unmap_phys_guard()` wants to know if this is about anonymous
mem or not which can probably be determined programmatically (whether there
is a corresponding page frame or backing store entry for it) but that is not
implemented now.

So to say that `MAP_FIXED` is not implementable at the moment due to those
details and it might make sense to simply return `ENOTSUP` for the time being.
